### ISSUE TITLE
Revisit public method/property/initializer signatures

### ISSUE THREAD

**Issue by allevato** [apple/swift-protobuf#12]
Several of the method/property/initializer signatures feel a bit un-Swifty and unclear about their purpose or their inputs/outputs.

For example, `init(protobuf:)` takes a `Data` argument but that's not clear from the argument label; likewise `serializeProtobuf()` returns `Data`. Off the top of my head, names that would better fit with [Swift's API naming guidelines](https://swift.org/documentation/api-design-guidelines/) would be `init(data:)` and `serializedData()`, as examples.

Nailing down best practices for API naming on the swift-evolution mailing list was a huge effort, and we should audit the user-facing APIs in SwiftProtobuf and ensure that they follow the same patterns for consistency.

**Comment by allevato** [comment]
Added note: It's also important that we do this very soon, to determine what changes we need to make before it gets too late to smoothly break things. (At least for methods/properties/initializers, we can keep the old ones around with a deprecation period).

**Comment by tbkka** [comment]
init(data:) and serializedData() are ambiguous, since they don't specify the format of the data.  It is possible, for instance, that we would someday provide support for NSKeyedArchiver serialization, which also produces and consumes a data object.

init(protobufData:) would satisfy both concerns.  I could live with that.

A similar problem would exist if you tried renaming init(json:) to init(string:).  That doesn't allow for the string being protobuf text vs. JSON.

**Comment by thomasvl** [comment]
I don't think anyone would suggest `init(string:)` because of the ambiguity within the protobuf space.  Playing devil's advocate, there are a few ways to express an object tree in JSON, but hopefully we don't need it to be `init(protobufJSON:)` since the protobuf context is implied.

A fair mount of the  [protocol buffer docs](https://developers.google.com/protocol-buffers/) (and other language apis), seem to just talk about _serialize_ being the action of creating a binary blob.  So within the protobuf space, it seems like it is likely defined.  Maybe `init(serialized:)`?

**Comment by allevato** [comment]
I would expect the string methods to be differentiated by something like `init(textString:)` and `init(jsonString:)`. But as we all know, naming things is the hardest part of software design. :)

The "natural" data representation of a protocol buffer message is its binary-coded wire format as defined by the spec, and that's what users of a protocol buffer implementation would expect. So, I don't find the `NSKeyedArchiver` use case very compelling, because this is an implementation of the Protocol Buffers specification in Swift, and `NSKeyedArchiver` is not part of that spec, so I would argue that it doesn't belong here. If someone wanted such a thing, they could add it in a new project or utility module (even as an extension to `ProtobufMessage`). And if we did go down that route, why wouldn't the methods take instances of `NSKeyedArchiver` (or even `NSCoder`) instead of the underlying data?

In other words, I don't want us to get into a situation where we complicate the fundamental/core API based on other possibilities that haven't been designed/vetted yet.

**Comment by tbkka** [comment]
`init(serialized:)` doesn't satisfy either of the concerns you and I discussed earlier:  It's ambiguous as to both runtime type and serialized format.

I don't have any strong preference between `init(jsonString:)` and `init(json:)`.  I think most people understand that JSON is a string format, but I could envision alternatives.

I agree that _in the context of protobuf_ `init(data:)` is clear.  But _in the context of my program that's using protobuf_ it might not be.  Even apart from the possibility of SwiftProtobuf supporting other codings, I've talked to plenty of people who do use multiple codings at different points in their applications.  So I would still argue that `init(protobufData:)` is the right way to go if you think `init(protobuf:)` is unclear.

**Comment by tbkka** [comment]
If you wanted to be fully type-accurate, I think the right signature would be:
```
    init(protobufSerializedData: ProtobufSerializedData)
```

where `ProtobufSerializedData` is just a simple wrapper around `Data`.  You could even decorate it with convenience methods such as:

```
struct ProtobufSerializedData {
   var data: Data
   init(_ data: Data)
   func deserialize<M: ProtobufMessage>(as message: M.Type) -> M
   init<M: ProtobufMessage>(protobufMessage: M)
}
```

With this, to deserialize a `Data` that you knew was protobuf, you would have to explicitly declare it as protobuf serialized data, using code such as the following:
```
  var m = MyMessage(protobufSerializedData: ProtobufSerializedData(data: myData))

  var m2 = ProtobufSerializedData(data: myData).deserialize(as: MyMessage.self)
```

In this model, note that `bytes` fields would _not_ be `ProtobufSerializedData`.  They would just be `Data`, I think.

I'm not necessarily proposing we do this; I just wanted to throw it out as a possibility for discussion.

**Comment by allevato** [comment]
I feel that that only adds noise without any extra safety, since a `Data` with any contents could be wrapped in a `ProtobufSerializedData` without any validation. If the programmer *knows* that a `Data` is protobuf, they should just be able to deserialize it without wrapping it. If they don't know if it's protobuf or not, then that's a problem.

I have to stick with my original feeling on this—whether or not someone uses multiple data encodings in an application, that shouldn't burden the protocol buffers API with excess verbosity. The protocol buffers project isn't a general purpose generator for any possible data format—it is the definition of a specific binary data representation (along with JSON and text) that provides the convenience of types generated in various languages. As such, there's no reason to require distinguishing "protobufData" from just "data", because the natural data representation is well-defined and unambiguous. Someone getting data in or out of a protobuf message shouldn't expect that data to be anything but PB wire format, and if they want that, they should be using something other than protobuf.

**Comment by allevato** [comment]
I've started doing some passes over the signatures in our public APIs, to see where names can be tweaked to better match Swift design guidelines. Here's a first pass, focusing on the core types that users will work with the most. Posting them here for questions/comments before I start working on a change.

# General changes

* `protoFieldNumber` -> `fieldNumber` throughout
* Improve generic argument names (e.g., `<T: FieldDecoder>` -> `<Decoder: FieldDecoder>`)

# Message.swift

## `protocol Message`

* `func serializeJSON()` -> `func serializedJSONString()`
* `func serializeAnyJSON()` -> `func serializedAnyJSONString()`

## `protocol Proto2Message`

* `var unknown` -> `var unknownFields`

# JSONTypeAdditions.swift

## `extension Message`

* `init(json: String)` -> `init(jsonString: String)`

# TextTypeAdditions.swift

## `extension Message`

* `init(text: String, extensions: ExtensionSet?)` -> `init(textString: String, extensions: ExtensionSet?)`
* `func serializeText()` -> `func serializedTextString()`

# ProtobufTypeAdditions.swift

## `extension Message`

The use of just "data" below without qualification emphasizes the fact that the binary encoding is the **natural** encoding of a message. Even if we add `Data`-based JSON serialization like we've talked about in the past, those can be qualified like `serializedJSONData` to emphasize the data format difference there.

* `func serializeProtobuf()` -> `func serializedData()`
* `func serializeProtobufSize()` -> `func serializedDataSize()`
* `init(protobuf: Data)` -> delete
* `init(protobuf: Data, extensions: ExtensionSet?)` -> `init(data: Data, extensions: ExtensionSet?)` with default `nil` for `extensions` to replace the above initializer

I believe the following two can be deleted--they only call `init` followed by `decodeIntoSelf`, so it should be possible to remove them, replace the one (?) place where it's actually used with the two operations mentioned previously:

* `init(protobufBytes: UnsafePointer<UInt8>, count: Int)`
* `init(protobufBytes: UnsafePointer<UInt8>, count: Int, extensions: ExtensionSet?)`

**Comment by thomasvl** [comment]
TextTypeAdditions.swift - The [C++](https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.text_format) and [Java](https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/TextFormat) call this TextFormat.  Seems like we should have names more inline with that?

I like _serializedData_ naming as it seems to be inline with the other languages in that they tend to use "serialize" for the binary encode.  The one thing you have I question is `init(data: Data, extensions: ExtensionSet?)`, seems like the first arg should be `serializedData` to maintain the naming consistency.

`serializedAnyJSONString` - what is this for, do we need this as a public api?

Why is `serializedJSONString` in Message.swift, but `init(jsonString: String)` in JSONTypeAdditions.swift?

**Comment by tbkka** [comment]
> protoFieldNumber -> fieldNumber throughout

This is already done in #243.

**Comment by tbkka** [comment]
> I believe the following two can be deleted...
>
> init(protobufBytes: UnsafePointer<UInt8>, count: Int)
> init(protobufBytes: UnsafePointer<UInt8>, count: Int, extensions: ExtensionSet?)

Good catch! #243 already has some progress on reducing the visible surface for the decoders, so I'll just fold this in there.

**Comment by tbkka** [comment]
> I believe the following two can be deleted...
> 
> init(protobufBytes: UnsafePointer, count: Int)
> init(protobufBytes: UnsafePointer, count: Int, extensions: ExtensionSet?)

This is now implemented in #243.

**Comment by allevato** [comment]
Agree; `textFormat` is more consistent with other implementations.

**Comment by allevato** [comment]
I think we have a solid set of names now—the symmetry between the argument labels for parsing initializers and the names of the serialize methods will be nice for end users.

Fixed by #276 and #282.
